site-src/geps/gep-724.md

```diff
@@ -0,0 +1,624 @@
+```go
+// ParentRef identifies an API object that should be considered a parent of this
+// resource. The only kind of parent resource with "Core" support is Gateway.
+// This API may be extended in the future to support additional kinds of parent
+// resources such as Routes.
+type ParentRef struct {
+  // Group is the group of the referent.
+  //
+  // Support: Core
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +kubebuilder:default=gateway.networking.k8s.io
+  Group string `json:"group"`
+
+  // Kind is kind of the referent.
+  //
+  // Support: Core (Gateway)
+  // Support: Extended (Other Resources)
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +kubebuilder:default=Gateway
+  Kind string `json:"kind"`
+
+  // Namespace is the namespace of the referent. When unspecified (empty
+  // string), this will either be:
+  //
+  // * local namespace of the target is a namespace scoped resource
+  // * no namespace (not applicable) if the target is cluster-scoped.
+  //
+  // Support: Extended
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +optional
+  Namespace string `json:"namespace,omitempty"`
+
+  // Scope represents if this refers to a cluster or namespace scoped resource.
+  // This may be set to "Cluster" or "Namespace".
+  //
+  // Support: Core (Namespace)
+  // Support: Extended (Cluster)
+  //
+	// +kubebuilder:validation:Enum=Cluster;Namespace
+  // +kubebuilder:default=Namespace
+  // +optional
+  Scope string `json:"scope,omitempty"`
+
+  // Name is the name of the referent.
+  //
+  // Support: Core
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  Name string `json:"name"`
+
+  // SectionName is the name of a section within the target resource. In the
+  // following resources, SectionName is interpreted as the following:
+  //
+  // * Gateway: Listener Name
+  //
+  // Implementations MAY choose to support attaching Routes to other resources.
+  // If that is the case, they MUST clearly document how SectionName is
+  // interpreted.
+  //
+  // When unspecified (empty string), this will reference the entire resource.
+  // For the purpose of status, an attachment is considered successful if at
+  // least one section in the parent resource accepts it. For example, Gateway
+  // listeners can restrict which Routes can bind to them by Route kind,
+  // namespace, or hostname. If 1 of 2 Gateway listeners accept attachment from
+  // the referencing Route, the Route MUST be considered successfully
+  // attached. If no Gateway listeners accept attachment from this Route, the
+  // Route MUST be considered detached from the Gateway.
+  //
+  // Support: Core
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +optional
+  SectionName string `json:"sectionName,omitempty"`
```

Comment:
`MinLength=0` or pointer type. 

I suggested on a call that we could use `ObjectFieldSelector` or similar to specify the section, but I neglected to comment on the PR earlier.  Would it make sense to use `ObjectFieldSelector` so the API is explicit about which section this references?

Reply:
Will update to a pointer, thanks!

As far as ObjectFieldSelector ([ref for anyone curious](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#objectfieldselector-v1-core)), I'm not sure it's quite what we're looking for here. As I understand it, it allows us to select a single field within a resource. That would allow for the valid `spec.listeners[1]`, but lots of invalid values as well (someone may select some other part of a Gateway that may not make any sense). 

Even the valid example of `spec.listeners[1]` has issues in that it would be tied to an index and not a name. That would require Gateway admins to maintain a very specific ordering of their listeners. If that order ever switched, lots of things could break. A similar argument could be made for changing a listener name, but that seems both more manageable and expected.

If I'm understanding the [relevant API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#selecting-fields) correctly, this field should be interpreted with standard JavaScript syntax, so I don't think we'd even get the flexibility of jsonpath here.

Reply:
What I had in mind was that `ParentRef` would have a field like `SectionFieldRef corev1.ObjectFieldSelector` in addition to the `SectionName string` field,  and the user would specify `sectionFieldRef: spec.listeners` along with `sectionName`.  The API could require that the field identified by `sectionFieldRef` be a `+listType=map`, and then `sectionName` would specify the value to match against the field identified by `+listMapKey`: 
```go
type GatewaySpec struct {
	// ...
	// +listType=map
	// +listMapKey=name
	Listeners []Listener `json:"listeners"`

	// ...
}
type ParentRef struct {
	// ...

	// SectionFieldRef specifies a field containing sections within the
	// referent.  The specified field must be listType=map.  If
	// SectionFieldRef is specified, then SectionName must be specified as
	// well, and the value of SectionName will be used to identify the
	// specific section by matching it against the field specified by
	// listTypeMap.
	//
	// Support: Core
	//
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=253
	// +optional
	SectionFieldRef corev1.ObjectFieldSelector `json:"sectionFieldRef,omitempty"`

	// SectionName is ...
	SectionName string `json:"sectionName,omitempty"`

	// ...
}
```
Example:
```yaml
kind: Gateway
metadata:
  namespace: infra
  name: lb
spec:
  listeners:
  - name: foo
    # ...
---
kind: HTTPRoute
# ...
spec:
  attachTo:
  - kind: Gateway
    namespace: infra
    name: lb
    sectionFieldRef.fieldPath: spec.listeners
    sectionName: foo
```

Reply:
`ObjectFieldSelector` might not be the best type (maybe it would be better to define a new  type that includes the field path and the item name), but the idea is to make explicit what `SectionName` is identifying.

Reply:
Thanks for the explanation @Miciah! That does make a lot more sense. I can see that being a natural extension point if/when we want to support attaching to other arbitrary fields in Gateway or elsewhere. If I'm understanding correctly, it looks like it would be easy to add as a new optional field in the future without any kind of breaking changes.